verifier: readability, quality, bugfixes

nalinbhardwaj opened the pull request:
<img width="650" alt="Screenshot 2023-09-08 at 4 01 45 PM" src="https://github.com/daimo-eth/eip-7212/assets/6984346/085e2e97-fa27-463a-9e56-6b3f7473f734">

nalinbhardwaj committed:
verifier: readability, quality, bugfixes

dcposch commented on src/P256Verifier.sol:
```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
-        return X;
+        if (scalar_u == 0 && scalar_v == 0) return (0, true);
+
+        (H0, H1, add_success) = ecAff_add(gx, gy, Q0, Q1);
```
```
// H = g + Q
```

nalinbhardwaj replied:
done

dcposch commented on src/P256Verifier.sol:
```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
+        }
+
+        // initialise (X, Y) depending on the first active bitpair.
+        if (bitpair == 0) { // if scalar_u = scalar_v = 0.
```
```
assert(bitpair > 0);
```

nalinbhardwaj replied:
added `//invariant` comments

dcposch commented on src/P256Verifier.sol:
```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
+        }
+        else if (bitpair == 1) {
+            X = gx;
+            Y = gy;
```
```
(X, Y) = (gx, gy);
```

nalinbhardwaj replied:
done

dcposch commented on src/P256Verifier.sol:
```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
+            Y = gy;
+        } else if (bitpair == 2) {
+            X = Q0;
+            Y = Q1;
```
might be worth calling this `(QX, QY)` or conversely `g0, g1`

nalinbhardwaj replied:
done

dcposch commented on src/P256Verifier.sol:
```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
+                T2 = H1;
+            }
+
+            // if (zz == 0) { // this should never happen?
```
assert

nalinbhardwaj replied:
add invariant

dcposch commented on src/P256Verifier.sol:
```diff
@@ -58,286 +58,189 @@ contract P256Verifier {
+        uint256 zz1;
+        uint256 zzz1;
 
-        if (ecAff_IsZero(x0, y0)) return (x1, y1);
```
!!!

dcposch commented on src/P256Verifier.sol:
```diff
@@ -347,107 +250,138 @@ contract P256Verifier {
+     * by Fermat's little theorem, compute u^(f-2) mod f using modexp precompile
+     * Assume f != 0.
+     */
+    function FCL_ModInv(uint256 u, uint256 f, uint256 minus_2modf) internal view returns (uint256 result, bool success) {
```
we can probably remove the FCL prefixes

nalinbhardwaj replied:
done

dcposch reviewed (commented):
Less gas
Less bugs
Less code

amazing 🚀

nalinbhardwaj committed:
review comments

dcposch committed:
doc: mit license copyright

dcposch committed:
test: Wycheproof vectors

- script downloads Wycheproof tests
- generates clean P256 verification vectors
- forge test reads that file, parses JSON
- tests P256Verifier

dcposch committed:
feat: ~all Wycheproof vectors, deduplicated

nalinbhardwaj committed:
fix wycheproof discovered bugs

nalinbhardwaj committed:
verifier: defensive r_x modulo

dcposch committed:
fix: secp256r1 + SHA-256 vectors only

dcposch committed:
test: validate extracted vectors using SubtleCrypto

dcposch committed:
fix: simplify vectors, "acceptable" > valid

dcposch committed:
test: add CI

dcposch committed:
test: add coverage

dcposch committed:
test: additionally validate vectors against @noble/curves

dcposch committed:
fix: ci forge build

dcposch committed:
fix: fix test, flag the 4 failing vectors

dcposch committed:
fix: assert precompiles
